tests/codex/testvalidation.nim

```diff
@@ -41,12 +56,170 @@ asyncchecksuite "validation":
+                              slots: HashSet): int =
+    if validationGroups.isNone:
+      return if maxSlots == 0: 100 else: maxSlots
+    
+    if maxSlots == 0 or slots.len < maxSlots:
+      return slots.len
+    else:
+      return maxSlots
+  
+  proc printExtraTestInfo(maxSlots: MaxSlots,
+                          validationGroups: ?ValidationGroups,
+                          groupIndex: uint16,
+                          slotsInSelectedGroup: int,
+                          expectedSlots: int) =
+    echo "---------------------------------------------------------" &
+            "---------------------------------------------------------"
+    if validationGroups.isNone:
+      echo fmt"       ⬇︎ {maxSlots=}, {validationGroups=}, " &
+          fmt"{groupIndex=}: {expectedSlots=} ⬇︎"
+    else:
+      echo fmt"       ⬇︎ {maxSlots=}, {validationGroups=}, " &
+          fmt"{groupIndex=}: {slotsInSelectedGroup=}, {expectedSlots=} ⬇︎"
+      
+
+  for (maxSlots, validationGroups, groupIndex) in [
+                                       (0, ValidationGroups.none, 0),
+                                       (0, ValidationGroups.none, 1),
+                                       (0, ValidationGroups(2).some, 0),
+                                       (0, ValidationGroups(2).some, 1),
+                                       (10, ValidationGroups.none, 0),
+                                       (10, ValidationGroups.none, 1),
+                                       (10, ValidationGroups(2).some, 0),
+                                       (10, ValidationGroups(2).some, 1)]:
+    test fmt"slots should be observed ({maxSlots=}, {validationGroups=}, " &
+        fmt"{groupIndex=})":
+      let validation = await setupAndStartValidation(
+        maxSlots, validationGroups, groupIndex.uint16)
+      var slotsInSelectedGroup = 0
+      var slots = initHashSet[SlotId]()
+      for i in 0..<100:
+        let slot = Slot.example
+        setupValidationGroup(slot.id, maxSlots, validationGroups,
+          groupIndex.uint16, slotsInSelectedGroup, slots)
+        await market.fillSlot(slot.request.id, slot.slotIndex,
+          proof, collateral)
+      await validation.stop()
+      let expectedSlots = calculateExpectedSlots(
+        maxSlots, validationGroups, slots)
+      printExtraTestInfo(maxSlots, validationGroups, groupIndex.uint16,
+                         slotsInSelectedGroup, expectedSlots)
+      check validation.slots.len == expectedSlots
+      check slots == toHashSet(validation.slots)
+
+  test "slot is not observed if it is not in the validation group":
+    let validationConfig = initValidationConfig(maxSlots, validationGroups,
+        (groupIndex + 1) mod uint16(!validationGroups))
+    let validation = Validation.new(clock, market, validationConfig)
+    await validation.start()
+    await market.fillSlot(slot.request.id, slot.slotIndex, proof, collateral)
+    await validation.stop()
+    check validation.slots.len == 0
```

Comment:
The main idea in these tests is to test that `fillSlot` causes the validator to add a slot to its list of slots to validate, right? The difference is that it is testing that sometimes slots are added and sometimes not, depending on the config. In my opinion, both of these two pieces of functionality (adding slots after `fillSlot`, and config dependent addition) are covered in isolation, and so I think maybe these tests aren't adding anything that isn't already covered in other tests.

Reply:
Yes, these tests test that some slots are sometimes added and sometimes not. The first one also tests that slots are effectively added to different groups and that the maxSlots parameter is working correctly in combination with other config options - this is not covered effectively by any other test (well, it is "kind-of" tested in some other tests, but those other tests are testing internals and are not really good tests, see my note above - in face of having the first bigger test above those other tests can and perhaps should be removed).

The second test that you mark above, is indeed somehow redundant in the face of the first test, but, is a bit easier to follow and therefore worth keeping as well - when things go wrong, having this test may help you to narrow down the issue slightly easier. Thus I would really prefer to keep them both (and especially the first one - this is the only test that makes me feel confident that the groups and maxSlots work well together).

Reply:
Actually, having this test, I would be inclined to close hide both `maxSlotsConstraintRespected`
`shouldValidateSlot` and remove the corresponding tests that are just testing internals and thus they do not really add much confidence here.

Reply:
> the maxSlots parameter is working correctly in combination with other config options - this is not covered effectively by any other test (well, it is "kind-of" tested in some other tests

If we change and keep the test above (https://github.com/codex-storage/nim-codex/pull/890/files#r1772573230) then I think the logic for maxSlots is covered well enough. The `onSlotFilled` event handler only checks `shouldValidateSlot`, so really we should make sure that function is covered for all parameter options. And I think it's pretty well covered but you could bolster with some edge case or boundary tests if you really wanted.

Reply:
I discussed this with Eric in a call. We agreed on the following:
- I will make `maxSlotsConstraintRespected` and `shouldValidateSlot`
- I will recheck the related tests to see if they are adding anything. What they cover is pretty much covered by other tests that do not rely on the internals. If there is anything missing I will cover that.

As much as I feel that the big test is very valuable, it does not seem to land, so I will remove it (and keep it for myself 😜.

Reply:
Done. Waiting on the last review round from Adam.